### PR TITLE
Support atom-wise and molecule-wise priors

### PR DESCRIPTION
**Pull request by raimis**
Implement the support of atom-wise (e.g. atomic reference energies) and molecule-wise (e.g. electrostatic term) priors.

- [x] Refactor the priors
- [x] Implement pre-reduce and post-reduce methods

### CODE REVIEW

**Comment by peastman** [comment]
This conflicts with #134, which implements the same feature in a different way.  Can we agree on which API we want to use?  I followed what @PhilippThoelke suggested.  If we decide to use the other API, let me change it there so we don't create conflicts.

**Comment by raimis** [comment]
@stefdoerr @PhilippThoelke what do you think?

**Comment by PhilippThoelke** [comment]
I think both #134 and this PR will be useful and as far as I understand it they're not mutually exclusive? @peastman or is the conflict more substantial?

**Comment by peastman** [comment]
> I think both https://github.com/torchmd/torchmd-net/pull/134 and this PR will be useful and as far as I understand it they're not mutually exclusive?

We just need to pick which API we prefer for the priors.  Either one can work for my needs.  I don't think it would make sense to do both, since they're basically alternate APIs for doing the same thing.

**Comment by PhilippThoelke** [comment]
I would say the API in this PR is more general and therefore preferable but I think it's still useful to keep e.g. the distinction between `prior_args` and `prior_init_args` from #134.

File: torchmdnet/priors/atomref.py [hunk 1]
```diff
@@ -1,54 +1,17 @@
-from abc import abstractmethod, ABCMeta
+from .base import BasePrior
```
Comment (stefdoerr) [inline torchmdnet/priors/atomref.py:1]: Use absolute instead of local imports. They mess up some testing frameworks

**Comment by stefdoerr** [comment]
I agree with Philipp, this implementation seems more generic but should be trivial to merge with the additional changes in #134

**Review by stefdoerr — APPROVED** [review]

**Pull request merged**